interfaces/builtin/dotfiles.go

```diff
@@ -0,0 +1,153 @@
+
+const dotfilesSummary = `allows access to files or directories`
+
+const dotfilesBaseDeclarationSlots = `
+  dotfiles:
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: true
+`
+
+const dotfilesConnectedPlugAppArmor = `
+# Description: Can access specific files or directories.
+# This is restricted because it gives file access to arbitrary locations.
+`
+
+type dotfilesInterface struct {
+	commonInterface
+}
+
+func validatePaths(attrName string, paths []interface{}) error {
+	for _, npp := range paths {
+		np, ok := npp.(string)
+		if !ok {
+			return fmt.Errorf("%q must be a list of strings", attrName)
+		}
+		if strings.HasSuffix(np, "/") {
+			return fmt.Errorf(`%q cannot end with "/"`, np)
+		}
+		if !strings.HasPrefix(np, "/") && !strings.HasPrefix(np, "$HOME/") {
+			return fmt.Errorf(`%q must start with "/" or "$HOME"`, np)
+		}
+		p := filepath.Clean(np)
+		if p != np {
+			return fmt.Errorf("%q must be clean", np)
+		}
+		if strings.Contains(p, "..") {
+			return fmt.Errorf(`%q contains invalid ".."`, p)
+		}
+		if strings.Contains(p, "~") {
+			return fmt.Errorf(`%q contains invalid "~"`, p)
+		}
+		if err := apparmor.ValidateFreeFromAARE(p); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (iface *dotfilesInterface) BeforePreparePlug(plug *snap.PlugInfo) error {
+	hasValidAttr := false
+	for _, att := range []string{"read", "write"} {
+		if _, ok := plug.Attrs[att]; !ok {
+			continue
+		}
+		il, ok := plug.Attrs[att].([]interface{})
+		if !ok {
+			return fmt.Errorf("cannot add dotfiles plug: %q must be a list of strings", att)
+		}
+		if err := validatePaths(att, il); err != nil {
+			return fmt.Errorf("cannot add dotfiles plug: %s", err)
+		}
+		hasValidAttr = true
+	}
+	if !hasValidAttr {
+		return fmt.Errorf(`cannot add dotfiles plug: needs valid "read" or "write" attribute`)
+	}
+
+	return nil
+}
+
+func formatPath(ip interface{}) (string, error) {
+	p, ok := ip.(string)
+	if !ok {
+		return "", fmt.Errorf("%[1]v (%[1]T) is not a string", ip)
+	}
+	prefix := ""
+	if strings.Count(p, "$HOME") > 0 {
+		p = strings.Replace(p, "$HOME", "@{HOME}", 1)
+		prefix = "owner "
+	}
+	p += "{,/,/**}"
+
+	return fmt.Sprintf("%s%q", prefix, filepath.Clean(p)), nil
+}
+
+func addSnippet(spec *apparmor.Specification, perm string, paths []interface{}) error {
+	for _, rp := range paths {
+		p, err := formatPath(rp)
+		if err != nil {
+			return err
+		}
+		spec.AddSnippet(fmt.Sprintf("%s %s", p, perm))
+	}
+	return nil
+}
+
+func (iface *dotfilesInterface) AppArmorConnectedPlug(spec *apparmor.Specification, plug *interfaces.ConnectedPlug, slot *interfaces.ConnectedSlot) error {
+	var reads, writes []interface{}
+	_ = plug.Attr("read", &reads)
+	_ = plug.Attr("write", &writes)
+
+	errPrefix := fmt.Sprintf(`cannot connect plug %s: `, plug.Name())
+	spec.AddSnippet(dotfilesConnectedPlugAppArmor)
+	if err := addSnippet(spec, "rk,", reads); err != nil {
+		return fmt.Errorf(errPrefix+"%v", err)
```

Comment:
Nitpick, errPrefix does not contain any meta-characters interpreted by the standard library formatting system but perhaps we ought to be double plus careful and instead do the more correct: `return fmt.Errorf("%s%v", errPrefix, err)`?